opsmngr/host.go

```diff
@@ -0,0 +1,182 @@
+)
+
+// HostService is an interface for interfacing with Hosts in MongoDB Ops Manager APIs
+type HostService interface {
+	Get(context.Context, string, string) (*Host, *atlas.Response, error)
+	GetByHostName(context.Context, string, string, int) (*Host, *atlas.Response, error)
+	List(context.Context, string, *HostListOptions) (*Hosts, *atlas.Response, error)
+	Monitoring(context.Context, string, *Host) (*Host, *atlas.Response, error)
+	UpdateMonitoring(context.Context, string, string, *Host) (*Host, *atlas.Response, error)
+	StopMonitoring(context.Context, string, string) (*atlas.Response, error)
+}
+
+type HostServiceOp struct {
+	Client atlas.RequestDoer
+}
+
+type Host struct {
+	Aliases            []string      `json:"aliases,omitempty"`
+	AuthMechanismName  string        `json:"authMechanismName,omitempty"`
+	ClusterID          string        `json:"clusterId,omitempty"`
+	Created            string        `json:"created,omitempty"`
+	GroupID            string        `json:"groupId,omitempty"`
+	Hostname           string        `json:"hostname"`
+	ID                 string        `json:"id,omitempty"`
+	IPAddress          string        `json:"ipAddress,omitempty"`
+	LastPing           string        `json:"lastPing,omitempty"`
+	LastRestart        string        `json:"lastRestart,omitempty"`
+	ReplicaSetName     string        `json:"replicaSetName,omitempty"`
+	ReplicaStateName   string        `json:"replicaStateName,omitempty"`
+	ShardName          string        `json:"shardName,omitempty"`
+	TypeName           string        `json:"typeName,omitempty"`
+	Version            string        `json:"version,omitempty"`
+	Username           string        `json:"username,omitempty"`
+	Password           string        `json:"password,omitempty"`
+	AlertsEnabled      bool          `json:"alertsEnabled,omitempty"`
+	Deactivated        bool          `json:"deactivated,omitempty"`
+	HasStartupWarnings bool          `json:"hasStartupWarnings,omitempty"`
+	Hidden             bool          `json:"hidden,omitempty"`
+	HiddenSecondary    bool          `json:"hiddenSecondary,omitempty"`
+	HostEnabled        bool          `json:"hostEnabled,omitempty"`
+	JournalingEnabled  bool          `json:"journalingEnabled,omitempty"`
+	LogsEnabled        bool          `json:"logsEnabled,omitempty"`
+	LowUlimit          bool          `json:"lowUlimit,omitempty"`
+	MuninEnabled       bool          `json:"muninEnabled,omitempty"`
+	ProfilerEnabled    bool          `json:"profilerEnabled,omitempty"`
+	SSLEnabled         bool          `json:"sslEnabled,omitempty"`
+	LastDataSizeBytes  float64       `json:"lastDataSizeBytes,omitempty"`
+	LastIndexSizeBytes float64       `json:"lastIndexSizeBytes,omitempty"`
+	MuninPort          int32         `json:"muninPort,omitempty"`
+	Port               int32         `json:"port"`
+	SlaveDelaySec      int64         `json:"slaveDelaySec,omitempty"`
+	UptimeMsec         int64         `json:"uptimeMsec,omitempty"`
```

Comment:
[q] false and zero(0) are omitted values, is this ok for this API? I see you are using it for a POST and a PATH so we need to be careful about it

Reply:
(We only care about list and get right now but checking this anyway)

Reply:
I have tested the post and path and it works. What is your idea? Making them pointers?

Reply:
The request body parameters for the update are:
**alertsEnabled**, **authMechanismName**, **logsEnabled**, **muninPort**, **password**, **profilerEnable**, **sslEnabled** and **username**. 

I tried changing **profilerEnable** and **muninPort** to false and zero and they did not change. I have also tried to change the **municPort** to 10101 and It worked. 
How do you suggest to fix this issue? I honestly do not know why these two values are omitted 🤕

Reply:
like I said at the beginning, false and zero are not serialized when `omitempty` because go considers false and zero as empty, we either use pointers to avoid this or remove the `omitempty`
only remove the `omitempty` if we don't do a patch, since again go would default boolean and ints to false and zero, so in a struct where you don't set them you'll be serialising the defaults and risk overwriting values, this is one of the reasons the atlas client has so many intergers an booleans as pointers

Reply:
I made them pointers and it works 👍

Reply:
set `profilerEnabled` and `sslEnabled` to `false` and `muninPort` to `0`

```
{
	"authMechanismName": "NONE",
	"created": "2020-04-02T10:57:39Z",
	"groupId": "5e85a9e4fbe92e1f6543af81",
	"hostname": "andreaMac",
	"id": "4c98455092ddaa239d0e3bc2f793745a",
	"typeName": "NO_DATA",
	"alertsEnabled": false,
	"hostEnabled": true,
	"logsEnabled": false,
	"profilerEnabled": false,
	"sslEnabled": false,
	"muninPort": 0,
	"port": 27117,
	"links": [
		{
			"rel": "self",
			"href": "http://localhost:9080/api/public/v1.0/groups/5e85a9e4fbe92e1f6543af81/hosts/4c98455092ddaa239d0e3bc2f793745a"
		},
		{
			"rel": "http://mms.mongodb.com/group",
			"href": "http://localhost:9080/api/public/v1.0/groups/5e85a9e4fbe92e1f6543af81"
		}
	]
}
```